### PR TITLE
Settings UI: update Akismet settings card with a field to enter API key

### LINKED ISSUE #6206: Settings: Security > Antispam; API Key field required (and remove other option)
We don't need the "show number of approved" option here at all, just enable it and remove the UI for it.

screen shot 2017-01-27 at 2 11 37 pm

What we _do_ need though, is a field where you can access/change your API key.

Normally I'd say that it should be behind an "Advanced Options" link, but since it will be the only thing in this card now, it can probably just be there?

### LINKED ISSUE #6207: Settings: Security > Antispam entire card needs to be conditional on Akismet
This entire card should be conditional upon me having Akismet installed/active, or handle if it's not.

If we're using banners to handle that, then this card should probably not appear until Akismet is installed/active.

Probably pending seeing the banners in action, and exactly how they will interact with these sorts of settings/cards.

### PR DESCRIPTION
**Pull request by eliorivero**
Saving/fetching data requires #6693 to be merged. This one handles the UI.

Fixes #6206
Fixes #6207

captura de pantalla 2017-03-21 a las 16 31 35

captura de pantalla 2017-03-21 a las 16 26 22

captura de pantalla 2017-03-21 a las 16 27 13

captura de pantalla 2017-03-21 a las 16 27 01

captura de pantalla 2017-03-21 a las 16 27 06

#### Changes proposed in this Pull Request:

* card is hidden if Akismet is inactive
* gone is the checkbox for number of approved comments
* introduces input field to add the API key
* balances the margins of SettingsGroup when it's inside FoldableCard

#### Testing instructions:

* You should see the API key field. If you rebase `add/akismet-save-key`

### CODE REVIEW

File: _inc/client/security/antispam.jsx [hunk 1]
```diff
@@ -3,36 +3,54 @@
...
+							<FormFieldset>
+								<FormLabel>
+									<span className="jp-form-label-wide">{ __( 'Your API key' ) }</span>
+									<TextInput
+										name={ 'wordpress_api_key' }
+										value={ this.props.getOptionValue( 'wordpress_api_key' ) }
+										disabled={ this.props.isSavingAnyOption( 'wordpress_api_key' ) }
+										onChange={ this.props.onOptionChange } />
+								</FormLabel>
+								<p className="jp-form-setting-explanation" >
```
Comment (georgestephanis) [inline _inc/client/security/antispam.jsx]: Would it be a good idea to maybe hide this if they already have a valid api key?

**Comment by dereksmart** [comment]
This is what we're doing on the dashboard with an invalid key (no key)

Maybe we just do something like this for the card? 

screen shot 2017-03-20 at 10 12 16 am

**Comment by eliorivero** [comment]
Regarding the text, I followed the layout and text that @ashleighaxios proposed for Calypso https://github.com/Automattic/wp-calypso/issues/12195
While I do agree that it's a bit confusing, for this pass we can go with it and tweak it later.

Nice catch about removing the key, and yeah, when the key is invalid, we should do something. Let's do the same than the dash item card you pointed.

**Comment by eliorivero** [comment]
> Users that don't know what an API Key is and don't really need to know -- it overcomplicates things for them.

Not sure if this is a concern for now. The field is hidden in a foldable card, so it's hidden unless user opens it. Even if they open it, to edit it, they still have to intentionally enter text and click the saving button.

However, one thing you mentioned 
> "Your API key is valid!"
makes me think if we should add something, like a checkmark ✅, that provides feedback to user that the current key in the field is valid. What do you think @ashleighaxios?

**Comment by dereksmart** [comment]
> makes me think if we should add something, like a checkmark

Maybe we use a compact green notice like we do for backups? 
screen shot 2017-03-20 at 10 31 14 am

**Comment by eliorivero** [comment]
Nice, that would be consistent with the security scanning card.

**Comment by georgestephanis** [comment]
Ah, I missed that it was a foldable card.

I'd still like the blurb afterwards to be contextually relevant to the current api key -- cc: @michelleweber or anyone from editorial for suggestions / direction?

**Comment by jeffgolenski** [comment]
@eliorivero per your request, and with the gracious help of @ashleighaxios here's a mockup of how this section should appear. 

screen shot 2017-03-20 at 12 37 28 pm

We want to ensure we're matched up with calypso in how we handle form validation messages. Also, let's use the label "antispam API key" (swapping 'antispam' if we're using a different name for the feature, of course) where we can—to prevent any possible API confusion within wp-admin. Sure the form lives in the spam filtering card, but like you said. better safe than sorry!

**Comment by georgestephanis** [comment]
Should antispam be capitalized or not?  The mockup and your comment seem to vary in usage.

**Comment by jeffgolenski** [comment]
@georgestephanis Fixed! It's a feature name. Might as well capitalize it!

**Comment by eliorivero** [comment]
@jeffgolenski just noticed you dropped a link. Where should the "Learn more" go to? Shouldn't it have period at the end?

**Comment by eliorivero** [comment]
No worries, will point it to `https://docs.akismet.com/getting-started/api-key/`

**Comment by eliorivero** [comment]
Updated:

Displays messages when key is valid and invalid

captura de pantalla 2017-03-20 a las 18 26 14

captura de pantalla 2017-03-20 a las 18 26 55

**Comment by eliorivero** [comment]
Updates:

- hide valid/invalid message when text input is cleared
- add new endpoint /module/<service>/key/check to check the validity of an API key for a service. Introduces methods to check Akismet API key.
- add actions/reducers to check the validity of an Akismet API key
- add new `QueryAkismetKeyCheck` component 
- add new endpoint supporting POST method that receives a key to check against a whitelisted service. Introduced methods to check against Akismet.
- allow to pass a key to check against a service on the fly as the user types it.
- when key is first retrieved block input field and show message in it. When key check finishes display a success message if key is valid and an error otherwise. When user types the key, display message that it's being checked and show success or error message afterwards

**Comment by jeffgolenski** [comment]
@eliorivero 

Solid work here! I'm curious. If the API key has been inserted and its valid. Can we hide the block of text that starts with "If you don't already have an API key get on here..."?

**Comment by eliorivero** [comment]
Sure @jeffgolenski 👍

**Comment by eliorivero** [comment]
Updated:

- only show explanation message under input field when key is invalid or field is empty.

Screenshots in PR description were updated.

**Comment by ashleighaxios** [comment]
Funny little block, but you made it work great. Ty

**Comment by eliorivero** [comment]
@georgestephanis @dereksmart I'm worried that someone could use the endpoint to guess keys. Should we limit the number of key checks during a minute?
The UI interface has a debounce of half second, but the endpoint doesn't, and maybe we should add some protection. Then again, not sure if I'm being overzealous.

**Comment by dereksmart** [comment]
@eliorivero Why not just use a more strict permissions check?  Is there a reason we need to make the $_GET request available to anyone who has access to the admin?  Why not limit to `manage_plugin` capability or something

**Comment by eliorivero** [comment]
Updates:

- check that user can `manage_options` (admin role) to use the API check endpoint.

**Comment by georgestephanis** [comment]
"guided through the process of getting one in a new window" -- is that overly specific?

Could this be confusing if it's actually a new tab?

Am I overthinking this?

Probably.

**Comment by dereksmart** [comment]
@eliorivero think it makes sense to disable the save button when the API key is rendered `invalid`?  Since saving won't work anyway...

screen shot 2017-03-22 at 10 12 50 am

**Comment by dereksmart** [comment]
When there is no API key, the UI still says that my site is protected from Spam. 

screen shot 2017-03-22 at 10 15 19 am

**Comment by dereksmart** [comment]
> "guided through the process of getting one in a new window" -- is that overly specific?

Well no, considering that currently the link in the card opens in the same window :)

**Comment by dereksmart** [comment]
Also, if the site has a paid plan, they simply could get the key re-activated just by going to the setup flow for akismet: `wordpress.com/plugins/setup/dereksmart.wpsandbox.me?only=akismet`

We should be doing this for plan accounts, but take them to the normal flow for those with akismet active without a plan.

**Comment by eliorivero** [comment]
Updated:

- don't enable the Save settings button if the key entered is invalid
- update foldable card text based on API key validity
- removed "in a new window" since we open in the same tab

**Comment by dereksmart** [comment]
Looks great elio! send the 🐑

**Comment by eliorivero** [comment]
@dereksmart I've addressed your remarks

- don't accept any whitespace.
- allow to save an empty key for Akismet key.
- allow checking for a key validity when the current key is invalid

**Comment by dereksmart** [comment]
This is working very nicely now.  Thank you @eliorivero for your hard work on this! 

🐑 !!!

**Pull request merged**